### PR TITLE
Add getThreadList using GraphQL

### PR DESCRIPTION
**Pull request by ravkr**
as old getThreadList stopped working (and now returns only empty array) just started making new version

there is separate field with id for group conversations and "one to one" conversations...
should I merge them to just one "threadID"?

### CODE REVIEW

**Comment by ravkr** [comment]
how about name of thread? When it is not set Facebook says it is `null`, should I replace it with an empty string? and what to do in case of  1 to 1 conversations? Put name of another person? an empty string? `null`?

**Comment by ravkr** [comment]
there is `"outgoing_bubble_color": "FF8C0077"` should we return `FF8C0077`, `8C0077` or `#8C0077`?

**Comment by Schmavery** [comment]
Thanks for working on this @ravkr 
The best guidance I can likely give is to look at the format* functions in utils.js and try to be consistent with those where possible.
formatMessage is probably one of the most stable, there we have these answers to your questions:
```js
threadID: formatID((originalMessage.thread_fbid || originalMessage.other_user_fbid).toString()),
threadName: (originalMessage.group_thread_info ? originalMessage.group_thread_info.name : originalMessage.sender_name),
```

(Of course, formatThread doesn't seem to follow the same convention but I think that's not by design and that the behavior in formatMessage is generally preferable)

As for the color, I think we have classically returned whatever facebook gives us, but I could see an argument for leaving out the initial `FF` (which is the alpha value, I guess)... not sure.

**Comment by Schmavery** [comment]
@ravkr are you still working on this PR or is it waiting for more review/merging? It would be really great if we could replace the broken getThreadList with a (fully?) compatible call made to graphql.

**Comment by ravkr** [comment]
I didn't have much time this week, sorry... but I will try to finish it on Sunday or Monday

**Comment by forcemagic** [comment]
You need some help with this @ravkr? I'm happy to assist if you have any problems :smiley:

**Comment by ravkr** [comment]
@forcemagic you can try to write documentation to that function
here is example how to use it https://pastebin.com/NAfZYiU5 - it will just dump current formatted data to a file (but it is not complete yet)

**Comment by ravkr** [comment]
can someone inspect what values can be in `rtc_call_data`?
i got only `NO_ONGOING_CALL` and `AUDIO_GROUP_CALL` but when it is not a group call I am also getting `NO_ONGOING_CALL`

_some examples from Facebook responses_

    "rtc_call_data": {
        "call_state": "NO_ONGOING_CALL",
        "server_info_data": "",
        "initiator": {
            "id": "100000000000000"
        }
    },

    "rtc_call_data": {
        "call_state": "NO_ONGOING_CALL",
        "server_info_data": "",
        "initiator": null
    },

    "rtc_call_data": {
        "call_state": "AUDIO_GROUP_CALL",
        "server_info_data": "GANsbGEYFkdST1VQOjIwMDAwMDAwMDAwMDAwMDAYCFlxZnhHU25nAA==",
        // decoded: "\u0018\u0003lla\u0018\u0016GROUP:2000000000000000\u0018\u0008YqfxGSng\u0000"
        "initiator": {
            "id": "100000000000000"
        }
    },

**Comment by Schmavery** [comment]
@ravkr the existing getThreadList doesn't have that field so it's fine to leave it out (unless you have a reason to really need it?)

**Comment by ravkr** [comment]
Thank you @forcemagic 👍 I've just merged it
@Schmavery I know that original getThreadList didn't have it, but why ignore something that Facebook send us? this new version returns few things more, but we can omit unknown data until we have a clue what it does 

there is one bigger change... now the functions returns `participants` - array with all users from conversation, along with info about them (this way calling `getUserInfo` for every userID is now obsolete)
of course function name is different... but should I map `participants` to make old `participantIDs` usable as before?

**Comment by forcemagic** [comment]
Happy to help @ravkr :) if I see any new changes I'll try to keep the docs up to date to reflect them.

**Comment by ravkr** [comment]
there is not much left to do....
obraz
(these are the currently ignored things)

**Comment by theweavrs** [comment]
@ravkr Is this completely implemented?

**Comment by Schmavery** [comment]
@ravkr what are the differences between what this call returns and what the original getThreadList returns? I would much rather prefer to replace the broken call than add a new one with a note saying the old one is broken

**Comment by ravkr** [comment]
@Schmavery these were removed by facebook: `serverTimestamp`, `isCanonical`, `isCanonicalUser`, `recipientsLoadable`, `hasEmailParticipant`, `readOnly`, `canReply`, `lastMessageType`
and now `participants` contains array of participants along with data about them (as Facebook sends it anyway), this is the reason for removing `participantIDs`... (you can still get IDs from `participants`)

any other fields should be compatible with the old version (correct me if I'm wrong)

**Comment by Schmavery** [comment]
If those are the only differences and we add a participantIDs field that is derived from participants, we can probably replace the broken function with this.
@bsansouci do you agree?

**Comment by forcemagic** [comment]
We could make an "alias" for every GraphQL query and make them backwards compatible. (Example: `getThreadList()` would actually call `getThreadListGraphQL()`. @Schmavery if you're okay with this, I would like to update the docs (and a bit of code), so that it reflects this approach.

**Comment by ravkr** [comment]
IMO `getThreadList()` should call `getThreadListGraphQL()` and print some warning message

**Comment by Schmavery** [comment]
Why do we need both?

**Comment by ravkr** [comment]
well... I'd put both for 2-3 months so people will get answer right in their console. otherwise some of them would probably create issues on github `why did you remove getThreadList?!?!? are there any alternatives!??!`

**Comment by Schmavery** [comment]
Except that as far as we can tell, getThreadList doesn't work. Anyway, we wouldn't remove getThreadList, we'd remove the test version suffixed with GraphQL.

**Comment by forcemagic** [comment]
@Schmavery We could extend `getThreadList()` with the new fields and somehow extract values for the old fields, so we don't need a new function, but we have already started creating `*GraphQL()` funcs, so... Who knows?

**Comment by Schmavery** [comment]
The plan was always to rename the existing *GraphQL functions (you'll notice they are not documented). Unfortunately they were written as prototypes/tests of using graphql so they need a little bit of compatibility work before they can be switched over. I'm just trying to save us that work now :)

**Comment by ritu99** [comment]
Does getThreadList() with graphQL need any more work? I can try and help out if needed.

**Comment by ravkr** [comment]
@ritu99 it is almost done but I do not have much time :( 
you can try to add support for different folders (look for `"tags":["INBOX"]` in javascript) and/or do some tests if everything works in "weird configurations" (talks with bots, banned users, disabled accounts, accounts waiting for verification, threads with people who blocked you etc.)

**Comment by ALiangLiang** [comment]
I discover two event types named "VideoCallMessage" and "VoiceCallMessage".

**Comment by ravkr** [comment]
from several messenger requests I've extracted these `tags`:
```
["ARCHIVED"]
["INBOX"]
["INBOX","unread"]
["OTHER","unread"]
["PENDING","unread"]
```

are there any other folders?

**Comment by ravkr** [comment]
I've just found two new account types `"UnavailableMessagingActor"` and `"ReducedMessagingActor"`

**Comment by Schmavery** [comment]
When this is ready, please rename it from getThreadListGraphQL to getThreadList and update the docs accordingly :)

**Comment by ravkr** [comment]
@Schmavery should I change `threadType: t.thread_type` to `isGroup:` with `true` or `false`?

somebody could also review the code, it should be almost complete now

**Comment by Schmavery** [comment]
As we did in with the other graphQL functions, please add `isGroup: bool` and remove `threadType` from the docs but keep `threadType` on the actual object (with the same semantics) marked with `// @Legacy` so we can remove it in the next major version update.  Will try to do more review when I get a chance.

**Comment by ravkr** [comment]
how about now? is everything okey?

**Comment by forcemagic** [comment]
For me, everything seems fine, let's see what @Schmavery says :smile:

**Comment by Budgetperson** [comment]
Quick comment as this as I was trying to use it early :). I think https://github.com/Schmavery/facebook-chat-api/pull/572/files#diff-017a6e261d15146e15fe740821e4aa47R104 needs to be changed to `isGroup: t.thread_type === "GROUP"`. The current implementation is causing an error for me. Thanks for the work on this PR!

**Comment by ravkr** [comment]
@Budgetperson thanks, fixed now

**Comment by ravkr** [comment]
support for `adminIDs` added

File: DOCS.md [hunk 1]
```diff
@@ -579,49 +579,255 @@ __Arguments__
...
+<td>Is the user verified? (Little blue tick mark)</td>
+</tr>
+<tr>
+<td>isMessageBlockedByViewer</td>
+<td>Is the user blocking messages from you?</td>
+</tr>
+<tr>
+<td>isViewerCoworker</td>
+<td>Is the user your coworker?
+<em>(not yet tested)</em>
```
Comment (Schmavery) [inline DOCS.md]: For these "not yet tested" ones, can we just leave them out of the docs if we don't know what they do?

Comment (Bjornskjald) [reply DOCS.md]: It's tested, it returns `true` if user is in the same company on workplace (see [here](https://github.com/ravkr/facebook-chat-api/files/1849633/workplace.json.txt))

**Comment by ravkr** [comment]
@Schmavery what do you think?

**Comment by Bjornskjald** [comment]
@Schmavery can you merge it? I'd like to use new getThreadList, but I'm not 100% sure about stability of @ravkr's repo (if it won't be deleted after merge)...

**Comment by Schmavery** [comment]
@ravkr thanks for working on this.

Let's not document untested fields (isEmployee, montageThread). We can add these later if someone understands and needs them.

I did a very quick comparison with the old getThreadList and see that we no longer have participantIDs. Can we add this and other missing fields to maintain compatibility? The same would probably apply between the format of the new participants field and the return value of getUserInfo, though I haven't had a chance to compare those myself yet.

There are several comments like `or maybe not? not tested` in the actual implementation. Can we get the signature of this function sorted out in general (and match the signature of the previous call, where we can)? Some things: make sure tags/type is optional, handle the case where it's a string, check the passed in type/tags values to make sure the user doesn't pass in something invalid.

Some of the logic around handling timestamp is not clear to me.

`esData[0].o0.data.viewer.message_threads.nodes` looks pretty brittle, are we sure this object will remain the same?

**Comment by Bjornskjald** [comment]
@Schmavery 
> see that we no longer have participantIDs

> now the functions returns participants - array with all users from conversation, along with info about them (this way calling getUserInfo for every userID is now obsolete)
of course function name is different... but should I map participants to make old participantIDs usable as before?
https://github.com/Schmavery/facebook-chat-api/pull/572#issuecomment-363220305

**Comment by Schmavery** [comment]
@Bjornskjald https://github.com/Schmavery/facebook-chat-api/pull/572#issuecomment-364579242

**Comment by Bjornskjald** [comment]
Uh, okay, haven't seen it...

**Comment by ravkr** [comment]
@Schmavery 
`participantIDs` should not be mentioned in the docs? Am I right?

these `or maybe not? not tested` are there because Facebook uses it, but we never can be sure if something else would work... :/  But I've just removed these comments

`resData[0].o0.data.viewer.message_threads.nodes` should work all the times
`resData[0]` is just the response (`[1]` is object with info about error/successful results)
`.o0.data` is result of our request (notice that we are sending `"o0"` in our `form` object)
etc..
obraz
It doesn't look like it would change anytime soon

I've explained logic around `timestamp` in the comments in code
```js
// When we ask for threads using timestamp from the previous request,
// we are getting the last thread repeated as the first thread in this response.
// .shift() gets rid of it
// It is also the reason for increasing limit by 1 when timestamp is set
// this way user asks for 10 threads, we are asking for 11,
// but after removing the duplicated one, it is again 10
```

**Comment by ravkr** [comment]
@Schmavery is everything okey now?

**Comment by delmotte** [comment]
when will this be available ? empty array is really boring...

**Comment by ravkr** [comment]
@Schmavery

**Comment by lgnap** [comment]
@Schmavery ?

**Comment by bsansouci** [comment]
My fault! I told him to not merge until I look at it. So looking at it today I promise ;)

File: src/getThreadList.js [hunk 2]
```diff
@@ -1,73 +1,211 @@
...
+function formatColor(color) {
+  if (color && color.match(/^(?:[0-9a-fA-F]{8})$/g)) {
+    return color.slice(2);
+  }
+  return color;
+}
 
-    if (end <= start) end = start + 20;
+function getThreadName(t) {
+  if (t.name || t.thread_key.thread_fbid) return t.name;
```
Comment (bsansouci) [inline src/getThreadList.js]: This looks weird, why do we return `t.name` even when it's null/undefined?

File: src/getThreadList.js [hunk 3]
```diff
@@ -1,73 +1,211 @@
...
-    };
+  for (let p of t.all_participants.nodes) {
+    if (p.messaging_actor.id === t.thread_key.other_user_id) return p.messaging_actor.name;
+  }
+}
 
-    form[type + "[offset]"] = start;
-    form[type + "[limit]"] = end - start;
+function mapNicknames(customizationInfo) {
+  return customizationInfo?customizationInfo.participant_customizations.map(u => {
```
Comment (bsansouci) [inline src/getThreadList.js]: can you do `customizationInfo && customizationInfo.participant_customizations ? ...`

File: src/getThreadList.js [hunk 4]
```diff
@@ -1,73 +1,211 @@
...
+    return {"userID": u.participant_id, "nickname": u.nickname};
+  }):[];
+}
 
-    if (ctx.globalOptions.pageID) {
-      form.request_user_id = ctx.globalOptions.pageID;
+function formatThreadList(data) {
+  return data.map(t => {
+    return {
+      threadID: utils.formatID(t.thread_key.thread_fbid || t.thread_key.other_user_id),
```
Comment (bsansouci) [inline src/getThreadList.js]: can you do `t.thread_key ? (t.thread_key.thread_fbid || t.thread_key.other_user_id) : null`

File: src/getThreadList.js [hunk 5]
```diff
@@ -1,73 +1,211 @@
...
+      // isCanonicalUser: t.is_canonical_neo_user, // is it always false?
+      // TODO: how about putting snippet in another object? current implementation does not handle every possibile message type etc.
+      snippet: t.last_message.nodes[0].snippet,
+      snippetAttachments: t.last_message.nodes[0].extensible_attachment, // TODO: not sure if it works
+      snippetSender: utils.formatID((t.last_message.nodes[0].message_sender.messaging_actor.id || "").toString()),
+      lastMessageTimestamp: t.last_message.nodes[0].timestamp_precise, // timestamp in miliseconds
+      lastReadTimestamp: (t.last_read_receipt.nodes[0]?t.last_read_receipt.nodes[0].timestamp_precise:null), // timestamp in miliseconds
+      cannotReplyReason: t.cannot_reply_reason, // TODO: inspect possible values
+
+      // @Legacy
```
Comment (bsansouci) [inline src/getThreadList.js]: <3 those annotations, thanks

File: src/getThreadList.js [hunk 6]
```diff
@@ -1,73 +1,211 @@
...
+      // can somebody inspect it? where is it used?
+      // probably Messenger Day uses it
+      reactionsMuteMode: t.reactions_mute_mode,
+      mentionsMuteMode: t.mentions_mute_mode,
+      isArchived: t.has_viewer_archived,
+      isSubscribed: t.is_viewer_subscribed,
+      timestamp: t.updated_time_precise, // in miliseconds
+      // isCanonicalUser: t.is_canonical_neo_user, // is it always false?
+      // TODO: how about putting snippet in another object? current implementation does not handle every possibile message type etc.
+      snippet: t.last_message.nodes[0].snippet,
```
Comment (bsansouci) [inline src/getThreadList.js]: Can we make this a bit more robust and check the presence of these elements before accessing? (same goes for everything below)

File: src/getThreadList.js [hunk 7]
```diff
@@ -1,73 +1,211 @@
...
+  });
+}
+
+module.exports = function(defaultFuncs, api, ctx) {
+  return function getThreadList(limit, timestamp, tags, callback) {
+    if (!callback && (utils.getType(tags) === "Function" || utils.getType(tags) === "AsyncFunction")) {
+      callback = tags;
+      tags = [""];
+    }
+    if (utils.getType(limit) !== "Number" || limit%1 !== 0 || limit <= 0) {
```
Comment (bsansouci) [inline src/getThreadList.js]: Can you do `isNaN` instead of this modulo trick?

Comment (bsansouci) [reply src/getThreadList.js]: Unless this isn't to check if limit is NaN?

File: src/getThreadList.js [hunk 8]
```diff
@@ -1,73 +1,211 @@
...
+      throw {error: "getThreadList: limit must be a positive integer"};
+    }
+    if (utils.getType(timestamp) !== "Null" &&
+       (utils.getType(timestamp) !== "Number" || timestamp%1 !== 0)) {
```
Comment (bsansouci) [inline src/getThreadList.js]: Same as above?

File: src/getThreadList.js (hunk 4 above)
Comment (ravkr) [reply src/getThreadList.js]: Why is it needed? Are there any cases where thread id wouldn't exist?

File: src/getThreadList.js (hunk 6 above)
Comment (ravkr) [reply src/getThreadList.js]: For everything? :/
Why not only for these with nodes[0]?

File: src/getThreadList.js (hunk 7 above)
Comment (ravkr) [reply src/getThreadList.js]: This is to check if the number is an integer and not a floating point number

File: src/getThreadList.js (hunk 8 above)
Comment (ravkr) [reply src/getThreadList.js]: Yup, it also looks for an integer

File: src/getThreadList.js (hunk 7 above)
Comment (forcemagic) [reply src/getThreadList.js]: How about `Number.isInteger()`? (See [mdn](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isInteger))

Comment (forcemagic) [reply src/getThreadList.js]: How about `Number.isInteger()`? (See [mdn](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isInteger))

File: src/getThreadList.js (hunk 4 above)
Comment (ravkr) [reply src/getThreadList.js]: AFAIK on Facebook almost everything have its ID 😢 
but I'll add it

File: src/getThreadList.js (hunk 7 above)
Comment (ravkr) [reply src/getThreadList.js]: good idea

File: src/getThreadList.js (hunk 6 above)
Comment (ravkr) [reply src/getThreadList.js]: done

**Comment by ravkr** [comment]
I think we can merge it now

**Comment by ravkr** [comment]
@bsansouci @Schmavery ?

**Comment by ravkr** [comment]
*two months later*

**Comment by bsansouci** [comment]
Done! Thanks for making this and keeping it up :)

**Pull request merged**